src/router_node.c

```diff
@@ -349,8 +349,17 @@ static bool AMQP_rx_handler(void* context, qd_link_t *link)
             next_delivery = pn_link_current(pn_link) != 0;
 
             uint64_t local_disp = qdr_delivery_disposition(delivery);
-            if (local_disp != 0) {
-                pn_delivery_update(pnd, local_disp);
+            //
+            // Call pn_delivery_update only if the local disposition is different than the pn_delivery's local disposition.
+            // This will make sure we call pn_delivery_update only when necessary.
+            //
+            if (local_disp != 0 && local_disp != pn_delivery_local_state(pnd)) {
+                //
+                // DISPATCH-1626 - This enables pn_delivery_update() and pn_delivery_settle() to be called back to back in the same function call.
+                // CORE_delivery_update() will handle most of the other cases where we need to call pn_delivery_update() followed by pn_delivery_settle().
+                //
+                if (qd_message_is_discard(msg))
```

Comment:
If the only point where the pn_delivery_update() needs to be done is if qd_message_is_discard(), then why not move all this dispo logic into the qd_message_is_discard(msg) logic right before the qd_delivery_settle() is done (linke #383)